### ISSUE TITLE
New Rule: no-compare-to-new

### ISSUE THREAD

**Issue by captbaritone** [eslint/eslint#15222]
### Rule details

Warns when attempting to compare objects/arrays by value

### Related ECMAScript feature

😶

### What type of rule is this?

Warns about a potential problem

### Example code

```js
function isEmpty(arr) {
  return arr === [];
}

function isDefault(obj) {
  return obj === {value: 'default'};
}
```

### Why should this rule be in the core instead of a plugin?

I implemented this rule and it *uncovered 131 errors* in our large code base. To me this indicates that it’s a rule that could provide value generally, and would be particularly valuable as a teaching tool for those coming from compare-by-value languages.

### Participation

- [X] I am willing to submit a pull request to implement this rule.

### Additional comments

In languages where objects are compared by value, you can test if an object (or array) is empty by creating a new empty instance and comparing it to the object you wish to test. However, in JavaScript, where object equality is based on referential identity, this approach does not work.

A comparison like this might look sensible to a person coming to JavaScript from a compare-by-value language:

```
function isEmpty(arr) {
  return arr === [];
}
```

In addition to empty objects, users may try to compare an existing object to a newly created object/array in a given state:

```
function isDefault(obj) {
  return obj === {value: 'default'};
}
```

Obviously — to us — these comparisons are nonsensical in JavaScript. Luckily it’s fairly trivial to detect statically. I implemented this rule and it **uncovered 131 errors** in our large code base. To me this indicates that it’s a rule that could provide value generally, and would be particularly valuable as a teaching tool for those coming from compare-by-value languages.

## An Implementation

Here, for reference and as a potential starting point, is the implementation I wrote for our code base. 

```
const COMPARISON_OPERATORS = new Set(['===', '!==']);

module.exports = {
  meta: {
    messages: {
      noCompareToNew:
        'In JavaScript, {{constructionType}}s are compared by identity, so a strict comparison to a newly created {{constructionType}} will always be false.',
    },
  },
  create(context) {
    return {
      BinaryExpression(node) {
        if (!COMPARISON_OPERATORS.has(node.operator)) {
          return;
        }
        const constructionType =
          newConstructionType(node.left) ?? newConstructionType(node.right);

        if (constructionType != null) {
          context.report({
            node,
            messageId: 'noCompareToNew',
            data: {constructionType},
          });
        }
      },
    };
  },
};

function newConstructionType(node) {
  switch (node.type) {
    case 'NewExpression':
    case 'ObjectExpression':
      return 'object';
    case 'ArrayExpression':
      return 'array';
    case 'FunctionExpression':
    case 'ArrowFunctionExpression':
      return 'function';
    case 'JSXElement':
    case 'JSXFragment':
      return 'React element';
    case 'ClassExpression':
      return 'class';
    case 'Literal':
      return node.value instanceof RegExp ? 'regular expression' : null;
    default:
      return null;
  }
}
```
